hasher-matcher-actioner/hmalib/common/timebucketizer.py

```diff
@@ -0,0 +1,161 @@
+# Copyright (c) Facebook, Inc. and its affiliates. All Rights Reserved
+import csv
+import datetime
+import os
+import typing as t
+
+MAX_BUFFER_SIZE = 100
+SECONDS_PER_DAY = 86400
+SECONDS_PER_MINUTE = 60
+
+
+Self = t.TypeVar("Self", bound="CSViable")
+
+
+class CSViable:
+    """
+    Interface of methods that must be implemented to run TimeBucketizer. It is used to guarantee safe parsing data in and out of CSV files
+    """
+
+    def to_csv(self) -> t.List[t.Union[str, int]]:
+        raise NotImplementedError
+
+    @classmethod
+    def from_csv(cls: t.Type[Self], value: t.List[str]) -> Self:
+        raise NotImplementedError
+
+
+T = t.TypeVar("T", bound=CSViable)
+
+
+class TimeBucketizer(t.Generic[T]):
+
+    # TODO Make this generic, so we can have TimeBucket[Tuple[int, int]] or TimeBucket[MoreComplexObject].
+    def __init__(
+        self, bucket_width: datetime.timedelta, storage_path: str, type: str, id: str, buffer_size: int = MAX_BUFFER_SIZE
+    ):
+        """
+        Divides the day into 24h / bucket_width buckets. When add_record
+        is called, based on current time, appends record to current bucket.
+        Each bucket's data is stored as a serialized CSV file.
+        Note that there may be multiple instances of TimeBucket writing to the
+        same file system so add some uniqueness for this instance.
+
+        Say your storage path is /var/data/threatexchange/timebuckets
+        The file for current datetime (2022/02/08/20:49 UTC) for a bucket_width
+        of 5 minutes should be "/var/data/threatexchange/timebuckets/<type>/2022/02/08/20:45/<unique_id>.csv"
+
+        The first <type> allows you to use the same storage folder for
+        say hashes and matches.
+        The second <unique_id> allows you to have multiple instances running
+        (eg. multiple lambdas) and writing to the same file system.
+        """
+
+        if bucket_width.total_seconds() < 60 or bucket_width.total_seconds() % 60 != 0:
+            raise Exception("Please ensure timedelta is atleast a minute long.")
+
+        if SECONDS_PER_DAY % bucket_width.total_seconds() != 0:
+            raise Exception(
+                "Time Delta must be equally divisible into buckets based on a 24 hour clock"
+            )
+
+        self.bucket_width = bucket_width
+        self.start, self.end = self._calculate_bucket_endpoints(datetime.datetime.now(), bucket_width)
+        self.storage_path = storage_path
+        self.id = id
+        self.type = type
+
+        self.buffer_size = buffer_size
+        self.buffer: t.List[T] = []
+
+
+    def _calculate_bucket_endpoints(self, time: datetime.datetime, bucket_width: datetime.timedelta):
```

Comment:
As discussed offline, make this a static method. So that it can be reused by class and instance methods.